Automattic/jetpack
# Backup: implement preflight checks in Jetpack dashboard

Related https://github.com/Automattic/jetpack-backup-team/issues/394

The idea is to start implementing the changes added in https://github.com/Automattic/jetpack/pull/36032 and https://github.com/Automattic/jetpack/pull/36044, to ensure Jetpack is not asking for credentials if preflight checks are success.

It is important to note that we are managing this with a feature flag on a WPCOM API (`GET rewind/preflight`), so by default the feature flag will be disabled and https://github.com/Automattic/jetpack/pull/36044/files#diff-c080df763461339c889fe76958fce59788330b7dcd90d7d552629defb0830af0R8 so we to ensure the affected pages will still work as usual.

## Proposed changes:

* Add `QueryBackupPreflightStatus` component that we could use on any Jetpack plugin page to fetch the preflight status once.
* Update VaultPress Backup card on Jetpack dashboard, to prevent asking for credentials if preflight status is success.
* Update `ProStatus` component to reflect the VaultPress Backup card as `Active` when preflight status is success.
* Update recommendation pages to prevent asking for credentials if preflight status is success.

### Screenshots

| Page | Before | After |
|---|---|---|
| jetpack#/dashboard | ![CleanShot 2024-02-28 at 23 11 34@2x](https://github.com/Automattic/jetpack/assets/1488641/f2dfce77-292c-4043-8e4a-8e7de92fd13c) | ![CleanShot 2024-02-28 at 23 10 14@2x](https://github.com/Automattic/jetpack/assets/1488641/5490962a-5746-4362-b129-7c085d3018be) |
| jetpack#/recommendations/summary | ![CleanShot 2024-02-28 at 23 56 37@2x](https://github.com/Automattic/jetpack/assets/1488641/8559c449-1c00-42c4-ba65-fb626e31dd13) | ![CleanShot 2024-02-28 at 23 55 47@2x](https://github.com/Automattic/jetpack/assets/1488641/b6033c72-3dd1-4a9a-86cd-f12c5c2e8694) |

### Other information:


## Jetpack product discussion


https://github.com/Automattic/jetpack-backup-team/issues/394

## Does this pull request change what data or activity we track or use?


No

## Testing instructions:





### Initial steps
* Use this branch in a site with a Jetpack VaultPress Backup plan and no credentials configured
* Open your Network tab in the developer tools

### Regression test
We want to ensure that having this feature disabled makes it to work as usual (requesting credentials if the site has no credentials configured). To do that:
* Navigate to `Jetpack > Dashboard` in the wp-admin menu (`/wp-admin/admin.php?page=jetpack#/dashboard`)
* Ensure there is an API call to `/wp-json/jetpack/v4/site/backup/preflight`
* In the response, the `feature_enabled` should be `false`
* You should see that the VaultPress Backup card is asking for server credentials in that page.
  ![CleanShot 2024-02-29 at 13 21 05](https://github.com/Automattic/jetpack/assets/1488641/bdf4102c-ae8f-4a01-a04e-1628d68c1fca)
* Now visit the `Recommendations` tab. Probably you will be prompted with something else, and then you can click on `Not now`. It should redirect to `/wp-admin/admin.php?page=jetpack#/recommendations/summary`
  * Note: if you navigated to that page after visiting the previous step (Jetpack Dashboard), you should not see any new API call to the backup preflight API. If you try to access `/wp-admin/admin.php?page=jetpack#/recommendations/summary` directly from the browser (reloading the page), you should see a new API call to that endpoint.
* In the recommendations tab, you should see the `Enable one-click restores` option in the right.
  ![CleanShot 2024-02-29 at 13 25 47](https://github.com/Automattic/jetpack/assets/1488641/fd95d716-f69d-4e13-869e-89702c63b5d2)

### Test with feature enabled
* Point `public-api.wordpress.com` to your sandbox and ensure you have the latest changes (trunk)
* In your sandbox config file, add this instruction to enable the feature flag:
  `add_filter( 'jetpack_backup_preflight_checks_enabled', '__return_true' );`
* Ensure your preflight checks are success (if you don't have any results, you can try to enqueue the preflight checks for that site. More info on this here: D130452-code
* Navigate to `Jetpack > Dashboard` in the wp-admin menu (`/wp-admin/admin.php?page=jetpack#/dashboard`)
* Ensure there is an API call to `/wp-json/jetpack/v4/site/backup/preflight`
* In the response, the `feature_enabled` should be `true`
* You should see that the VaultPress Backup card is no longer asking for server credentials:
  ![CleanShot 2024-02-28 at 23 10 14@2x](https://github.com/Automattic/jetpack/assets/1488641/5490962a-5746-4362-b129-7c085d3018be)
* Now visit the `Recommendations` tab. Probably you will be prompted with something else, and then you can click on `Not now`. It should redirect to `/wp-admin/admin.php?page=jetpack#/recommendations/summary`
* In the recommendations tab, you should see the `Enable one-click restores` option in the right.
  ![CleanShot 2024-02-28 at 23 55 47@2x](https://github.com/Automattic/jetpack/assets/1488641/b6033c72-3dd1-4a9a-86cd-f12c5c2e8694)
